File: sample/v3.1.2/C#/Dail/Dial/CAS_Common_Cfg.cs
using System;
using System.Text;
using System.Runtime.InteropServices;

namespace DJKeygoe
{
    public class CAS_Common_Code
    {
        /*************************************************************************************
        return 
	        0:	OK.
	        -1:	Fail, m_u8CalledTableCount Invalid
	        -2: Fail, m_u8CalledLen Invalid
	        -3: Fail, m_u8CalledTimeOut Invalid
	        -4: Fail, m_u8AreaCodeLen Invalid
	        -5: Fail, m_CalledTable[x].m_u8NumLen Invalid
        *************************************************************************************/
        public static unsafe int CAS_Common_Cfg_ReadCfg(ref CmdParamData_CAS_t pParam_CAS)
        {
            ClsIniFile clsIniFile = new ClsIniFile("C:\\DJKeygoe\\Samples\\CAS_Common_Code\\XMS_CAS_Cfg.INI");

            //int			i;
            //char		TmpStr[32], TmpName[32];

            // ------------------------ [Rule] ------------------------
            // m_u8CalledTableCount
            int iTmp = clsIniFile.GetFileInt("Rule", "CalledTableCount", 0);
            if ((iTmp < 0) || (iTmp > 16))
                return -1;							// m_u8CalledTableCount Invalid
            pParam_CAS.m_u8CalledTableCount = (byte)iTmp;

            // m_u8CalledLen
            iTmp = clsIniFile.GetFileInt("Rule", "CalledLen", 5);
            if ((iTmp <= 0) || (iTmp > 32))
                return -2;							// m_u8CalledLen Invalid
            pParam_CAS.m_u8CalledLen = (byte)iTmp;

            // m_u8CalledTimeOut
            iTmp = clsIniFile.GetFileInt("Rule", "CalledTimeOut", 0);
            if ((iTmp < 0) || (iTmp > 10))
                return -3;							// m_u8CalledTimeOut Invalid
            pParam_CAS.m_u8CalledTimeOut = (byte)iTmp;

            // m_u8NeedCaller
            iTmp = clsIniFile.GetFileInt("Rule", "NeedCaller", 0);
            if (iTmp != 0)
                iTmp = 1;
            pParam_CAS.m_u8NeedCaller = (byte)iTmp;

            // m_u8AreaCodeLen
            iTmp = clsIniFile.GetFileInt("Rule", "AreaCodeLen", 0);
            if (pParam_CAS.m_u8NeedCaller != 0)		// need caller
            {
                if ((iTmp < 0) || (iTmp > 10))
                    return -4;						// m_u8AreaCodeLen Invalid
            }
            pParam_CAS.m_u8AreaCodeLen = (byte)iTmp;

            // m_u8ControlMode
            iTmp = clsIniFile.GetFileInt("Rule", "ControlMode", 0);
            if ((iTmp != 0) && (iTmp != 1) && (iTmp != 2))
                iTmp = 0;
            pParam_CAS.m_u8ControlMode = (byte)iTmp;

            // ------------------------ [CalledTable] ------------------------
            //memset ( pParam_CAS->m_CalledTable, 0, sizeof(DJDataDefClass.CAS_CalledTableDesc_t) * 16 );
            //fixed(CAS_CalledTableDesc_t* pCalledTable = &pParam_CAS.m_CalledTable[0])
            fixed (byte* p = pParam_CAS.m_CalledTable)
            {
                CAS_CalledTableDesc_t* pCalledTable = (CAS_CalledTableDesc_t*)p;
                for (int i = 0; i < pParam_CAS.m_u8CalledTableCount; ++i)
                {
                    // m_u8NumHeadLen & m_u8NumHead
                    string TmpName = "NumHead[" + i.ToString() + "]";

                    StringBuilder strBlderTemp = new StringBuilder(256);
                    clsIniFile.GetFileString("CalledTable", TmpName, "168", strBlderTemp, 256);
                    string strTemp = strBlderTemp.ToString();
                    int j = 0;
                    for (j = 0; j < strTemp.Length && j < 15; ++j)
                    {
                        pCalledTable[i].m_u8NumHead[j] = (byte)strTemp[j];
                    }
                    if (j >= 15)
                    {
                        j = 14;
                        pCalledTable[i].m_u8NumHead[j] = (byte)'\0';
                    }
                    else
                    {
                        pCalledTable[i].m_u8NumHead[j] = (byte)'\0';
                    }
                    pCalledTable[i].m_u8NumHeadLen = (byte)j;

                    // m_u8NumLen
                    string TmpNameEx = "NumLen[" + i.ToString() + "]";
                    iTmp = clsIniFile.GetFileInt("CalledTable", TmpNameEx, 5);
                    if ((iTmp < 0) || (iTmp > 16))
                        return -5;							// m_CalledTable[x].m_u8NumLen Invalid
                    pCalledTable[i].m_u8NumLen = (byte)iTmp;

                    TmpName = null;
                    strBlderTemp = null;
                    TmpNameEx = null;
                    strTemp = null;
                }
            }

            // ------------------------ Other ------------------------
            // Call Out Parameter, set to default value directly.
            pParam_CAS.m_u8KA = 1;
            pParam_CAS.m_u8KD = 3;

            return 0;		// OK
        }
    };
}

File: sample/v3.1.2/Samples/Delphi/XMS_Demo/CAS_Common_Cfg.pas
unit CAS_Common_Cfg;

interface
uses
  Windows,SysUtils,inifiles,DJAcsDataDef;

Var
  CAS_Cfg_IniName:string;
  function CAS_Common_Cfg_ReadCfg(Var g_Param_CAS:CmdParamData_CAS_t):integer;
implementation

function CAS_Common_Cfg_ReadCfg(Var g_Param_CAS:CmdParamData_CAS_t):integer;
Var
  TmpStr:string;
  TmpName:string;
  pStr:Pchar;
  iTmp,i,j:integer;
  inifile:Tinifile;
begin

    CAS_Cfg_IniName:='.\XMS_CAS_Cfg.INI';

      inifile:=TiniFile.Create(CAS_Cfg_IniName);
          iTmp:=inifile.ReadInteger('Rule','CalledTableCount',1);
           if (iTmp<0) or (iTmp>16) then
            begin
              result:=-1;
              exit;
            end
           else
            begin
              g_Param_CAS.m_u8CalledTableCount:=iTmp;
            end;

          iTmp:=inifile.ReadInteger('Rule','CalledLen',5);
           if (iTmp<=0) or (iTmp>32) then
            begin
              result:=-2;
              exit;
            end
          else
            begin
              g_Param_CAS.m_u8CalledLen:=BYTE(iTmp);
            end;

          iTmp:=inifile.ReadInteger('Rule','CalledTimeOut',0);
           if (iTmp<0) or (iTmp>10) then
            begin
              result:=-3;
              exit;
            end
           else
            begin
              g_Param_CAS.m_u8CalledTimeOut:=BYTE(iTmp);
            end;

          iTmp:=inifile.ReadInteger('Rule','NeedCaller',1);
           if iTmp<>0 then
            begin
              iTmp:=1;
            end;
          g_Param_CAS.m_u8NeedCaller:=BYTE(iTmp);

          iTmp:=inifile.ReadInteger('Rule','AreaCodeLen',0);
          if g_Param_CAS.m_u8NeedCaller <> 0 then		// need caller
	         begin
		          if (iTmp <0) or (iTmp > 10) then
                begin
                  result:=-4;
                  exit;
                end;
           end;
          g_Param_CAS.m_u8AreaCodeLen := BYTE(iTmp);

          iTmp:=inifile.ReadInteger('Rule','ControlMode',0);
	        if (iTmp <> 0) and (iTmp <> 1) and (iTmp <> 2) then
		        iTmp := 0;
	        g_Param_CAS.m_u8ControlMode := Byte(iTmp);

          //FillChar(g_gLobalVar.g_Param_CAS.m_CalledTable,sizeof(CAS_CalledTableDesc_t)*16,0);
          ZeroMemory(pointer(@g_Param_CAS.m_CalledTable),sizeof(CAS_CalledTableDesc_t)*16);
          if  g_Param_CAS.m_u8CalledTableCount>0 then
            begin
              for i:=0 to g_Param_CAS.m_u8CalledTableCount-1 do
                begin
                  TmpName:=Format('NumHead[%d]',[i]);
                  TmpStr:=inifile.ReadString('CalledTable',TmpName,'168');
                  pStr:=Pchar(TmpStr);
                  if strlen(pStr)>14 then
                   pStr[14]:=char(0);
                  g_Param_CAS.m_CalledTable[i].m_u8NumHeadLen:=strlen(pStr);

                  if g_Param_CAS.m_CalledTable[i].m_u8NumHeadLen>0 then
                  begin
                    for j:=0 to g_Param_CAS.m_CalledTable[i].m_u8NumHeadLen-1 do
                      g_Param_CAS.m_CalledTable[i].m_u8NumHead[j]:=Byte(pStr[j]);
                  end;

                  TmpName:=Format('NumLen[%d]',[i]);
                  iTmp:=inifile.ReadInteger('CalledTable',TmpName,5);
                  if (iTmp<0) or (iTmp>16) then
                    begin
                      result:=-5;
                      exit;
                    end;
                  g_Param_CAS.m_CalledTable[i].m_u8NumLen := Byte(iTmp);
                end;
            end;
      g_Param_CAS.m_u8KA := 1;
      g_Param_CAS.m_u8KD := 3;
      inifile.Free;
      result:=0;
end;

end.
